pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +70,118 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
 			unix.Close(eventfd)
 		}
 	}()
-	glog.V(3).Infof("eviction: setting notification threshold to %s", threshold)
-	config := fmt.Sprintf("%d %d %s", eventfd, watchfd, threshold)
+	epfd, err := unix.EpollCreate1(0)
+	if err != nil {
+		glog.Warningf("eviction manager: error creating epoll: %v", err)
+		return nil, err
+	}
+	if epfd < 0 {
+		err = fmt.Errorf("EpollCreate1 call failed")
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			unix.Close(epfd)
+		}
+	}()
+	config := fmt.Sprintf("%d %d %d", eventfd, watchfd, threshold)
 	_, err = unix.Write(controlfd, []byte(config))
 	if err != nil {
 		return nil, err
 	}
-	return &memcgThresholdNotifier{
-		watchfd:     watchfd,
-		controlfd:   controlfd,
-		eventfd:     eventfd,
-		handler:     handler,
-		description: description,
-	}, nil
+	notifier := &linuxCgroupNotifier{
+		watchfd:   watchfd,
+		controlfd: controlfd,
+		eventfd:   eventfd,
+		epfd:      epfd,
+		stop:      make(chan struct{}),
+	}
+	return notifier, nil
 }
 
-func getThresholdEvents(eventfd int, eventCh chan<- struct{}, stop ThresholdStopCh) {
+func (n *linuxCgroupNotifier) Start(eventCh chan<- struct{}) {
+	err := unix.EpollCtl(n.epfd, unix.EPOLL_CTL_ADD, n.eventfd, &unix.EpollEvent{
+		Fd:     int32(n.eventfd),
+		Events: unix.EPOLLIN,
+	})
+	if err != nil {
+		glog.Warningf("eviction manager: error adding epoll eventfd: %v", err)
+		return
+	}
 	for {
-		buf := make([]byte, 8)
-		_, err := unix.Read(eventfd, buf)
+		event, err := wait(n.epfd, n.eventfd)
 		if err != nil {
+			glog.Warningf("eviction manager: error while waiting for memcg events: %v", err)
 			return
+		} else if event {
```

Comment:
nitpick: Handle the case where event is false with a continue, that way you can unindent the main part of the code. Also, should we log/warn on that case? Or is it expected?

```golang
if err != nil {
  glog.Warningf(...)
  return
}
if !event {
  glog.Warningf(...)  // Should we warn in this case?
  continue
}
// Consume the event from the eventfd
```

Reply:
done.  We expect wait to return false most of the time, as we are setting a 10 sec timeout.